feature-assets/src/main/java/io/novafoundation/nova/feature_assets/presentation/transaction/detail/swap/SwapDetailViewModel.kt

```diff
@@ -0,0 +1,158 @@
+import io.novafoundation.nova.feature_account_api.data.mappers.mapChainToUi
+import io.novafoundation.nova.feature_account_api.presenatation.account.icon.createAccountAddressModel
+import io.novafoundation.nova.feature_account_api.presenatation.account.wallet.WalletUiUseCase
+import io.novafoundation.nova.feature_account_api.presenatation.actions.ExternalActions
+import io.novafoundation.nova.feature_account_api.presenatation.chain.icon
+import io.novafoundation.nova.feature_assets.R
+import io.novafoundation.nova.feature_assets.presentation.AssetsRouter
+import io.novafoundation.nova.feature_assets.presentation.model.ChainAssetWithAmountParcelModel
+import io.novafoundation.nova.feature_assets.presentation.model.OperationParcelizeModel
+import io.novafoundation.nova.feature_swap_api.domain.model.rateAgainst
+import io.novafoundation.nova.feature_swap_api.presentation.formatters.SwapRateFormatter
+import io.novafoundation.nova.feature_swap_api.presentation.view.SwapAssetView
+import io.novafoundation.nova.feature_swap_api.presentation.view.bottomSheet.description.launchSwapRateDescription
+import io.novafoundation.nova.feature_wallet_api.data.network.blockhain.types.Balance
+import io.novafoundation.nova.feature_wallet_api.domain.ArbitraryTokenUseCase
+import io.novafoundation.nova.feature_wallet_api.domain.model.TokenBase
+import io.novafoundation.nova.feature_wallet_api.domain.model.withAmount
+import io.novafoundation.nova.feature_wallet_api.presentation.model.AmountModel
+import io.novafoundation.nova.feature_wallet_api.presentation.model.AmountSign
+import io.novafoundation.nova.feature_wallet_api.presentation.model.fullChainAssetId
+import io.novafoundation.nova.feature_wallet_api.presentation.model.mapAmountToAmountModel
+import io.novafoundation.nova.runtime.multiNetwork.ChainRegistry
+import io.novafoundation.nova.runtime.multiNetwork.asset
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+import kotlin.time.Duration.Companion.milliseconds
+
+class SwapDetailViewModel(
+    private val router: AssetsRouter,
+    private val addressIconGenerator: AddressIconGenerator,
+    private val chainRegistry: ChainRegistry,
+    private val externalActions: ExternalActions.Presentation,
+    private val arbitraryTokenUseCase: ArbitraryTokenUseCase,
+    private val walletUiUseCase: WalletUiUseCase,
+    private val swapRateFormatter: SwapRateFormatter,
+    private val descriptionBottomSheetLauncher: DescriptionBottomSheetLauncher,
+    val operation: OperationParcelizeModel.Swap,
+) : BaseViewModel(),
+    ExternalActions by externalActions,
+    DescriptionBottomSheetLauncher by descriptionBottomSheetLauncher {
+
+    private val tokenIn = historicalTokenFlow(operation.amountIn)
+    private val tokenOut = historicalTokenFlow(operation.amountOut)
+    private val tokenFee = historicalTokenFlow(operation.amountFee)
```

Comment:
It seems to me that we can catch a crash here if we disable the asset for some reason in chains.json